File: test/parse_log.py
#!/usr/bin/env python3

from enum import Enum, auto
from pathlib import Path


def fix(line):
    return line.replace(", )", ")").replace(", ]", "]").replace(", }", "}")


def splitlines(line):
    delims = "]})"

    out_line = ""
    for c in line:
        out_line += c
        if c in delims:
            out_line += "\n"

    return out_line


def parse_metis(log):
    class State(Enum):
        PRE_METIS = auto()
        METIS = auto()

    metis_log = [""]
    extras = [""]

    state = State.PRE_METIS

    for line in log.splitlines():
        target = extras

        if state == State.PRE_METIS:
            if line.startswith("     Running `"):
                state = State.METIS
        elif state == State.METIS:
            target = metis_log
        else:
            assert False

        target[0] += splitlines(fix(line)) + "\n"

    return min(metis_log), min(extras)


def parse_smite(log):
    class State(Enum):
        PRE_SMITE = auto()
        SMITE = auto()

    smite_log = [""]
    extras = [""]

    state = State.PRE_SMITE

    for line in log.splitlines():
        target = extras

        if state == State.PRE_SMITE:
            if line.startswith("METIS RESULT: ["):
                state = State.SMITE
        elif state == State.SMITE:
            target = smite_log
        else:
            assert False

        target[0] += splitlines(line) + "\n"

    return min(smite_log), min(extras)


def main(metis_log_path, smite_log_path, extras_log_path):
    metis_log_path = Path(metis_log_path)
    smite_log_path = Path(smite_log_path)
    extras_log_path = Path(extras_log_path)

    metis_log = metis_log_path.read_text()
    smite_log = smite_log_path.read_text()

    # log_size = min(len(metis_log), len(smite_log)) + 1000000
    log_size = int(1e10)

    metis_log, metis_extras = parse_metis(metis_log[:log_size])
    smite_log, smite_extras = parse_smite(smite_log[:log_size])

    metis_log_path.write_text(metis_log)
    smite_log_path.write_text(smite_log)
    extras_log_path.write_text(metis_extras + smite_extras)


if __name__ == "__main__":
    import argh

    argh.dispatch_command(main)

File: test/parse_log.rs
#!/usr/bin/env scriptisto

// scriptisto-begin
// script_src: src/main.rs
// build_cmd: cargo build --release && strip ./target/release/script
// target_bin: ./target/release/script
// files:
//  - path: Cargo.toml
//    content: |
//     package = { name = "script", version = "0.1.0", edition = "2018"}
//     [dependencies]
//     clap = { version = "3.1.18", features = ["derive"] }
// scriptisto-end

use std::fs;
use clap::Parser;

fn fix(line: &str) -> String {
    line.replace(", )", ")")
        .replace(", ]", "]")
        .replace(", }", "}")
}

fn splitlines(line: &str) -> String {
    let delims = "]})";

    let mut out_line = String::new();
    for c in line.chars() {
        out_line.push(c);
        if delims.find(c).is_some() {
            out_line += "\n";
        }
    }

    out_line
}

fn parse_metis(log: &str) -> (String, String) {
    enum State {
        PreMetis,
        Metis,
    }

    let mut metis_log = String::new();
    let mut extras = String::new();

    let mut state = State::PreMetis;

    for line in log.lines() {
        let mut target: &mut String = &mut extras;

        match state {
            State::PreMetis => {
                if line.starts_with("     Running `") {
                    state = State::Metis;
                }
            }
            State::Metis => {
                target = &mut metis_log;
            }
        }

        target.to_owned().push_str(&splitlines(&fix(line)));
        target.push('\n');
    }

    (metis_log, extras)
}

fn parse_smite(log: &str) -> (String, String) {
    enum State {
        PreSmite,
        Smite,
    }

    let mut smite_log = String::new();
    let mut extras = String::new();

    let mut state = State::PreSmite;

    for line in log.lines() {
        let mut target: &mut String = &mut extras;

        match state {
            State::PreSmite => {
                if line.starts_with("METIS RESULT: [") {
                    state = State::Smite;
                }
            }
            State::Smite => {
                target = &mut smite_log;
            }
        }

        target.to_owned().push_str(&splitlines(&fix(line)));
        target.push('\n');
    }

    (smite_log, extras)
}

#[derive(Parser, Debug)]
#[clap(name = "script", about = "A script.")]
struct Args {
    metis_log_path: String,
    smite_log_path: String,
    extras_log_path: String,
}

fn main() -> std::io::Result<()> {
    let opt = Args::from_args();

    let metis_log = fs::read_to_string(&opt.metis_log_path)?;
    let smite_log = fs::read_to_string(&opt.smite_log_path)?;

    let (metis_log, metis_extras) = parse_metis(&metis_log);
    let (smite_log, smite_extras) = parse_smite(&smite_log);

    fs::write(&opt.metis_log_path, metis_log)?;
    fs::write(&opt.smite_log_path, smite_log)?;
    fs::write(&opt.extras_log_path, metis_extras + &smite_extras)?;

    Ok(())
}
